File: src/main/java/uk/co/unclealex/hammers/calendar/client/views/Calendar.java
/**
 * 
 */
package uk.co.unclealex.hammers.calendar.client.views;

import javax.inject.Inject;

import uk.co.unclealex.hammers.calendar.client.presenters.CalendarPresenter.Display;
import uk.co.unclealex.hammers.calendar.client.util.CanWaitSupport;

import com.google.gwt.core.client.GWT;
import com.google.gwt.uibinder.client.UiBinder;
import com.google.gwt.uibinder.client.UiFactory;
import com.google.gwt.uibinder.client.UiField;
import com.google.gwt.uibinder.client.UiTemplate;
import com.google.gwt.user.client.ui.Button;
import com.google.gwt.user.client.ui.CheckBox;
import com.google.gwt.user.client.ui.DialogBox;
import com.google.gwt.user.client.ui.ListBox;
import com.google.gwt.user.client.ui.PopupPanel;
import com.google.gwt.user.client.ui.SimplePanel;
import com.google.gwt.user.client.ui.Widget;

/**
 * Copyright 2011 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.    
 *
 * @author unclealex72
 *
 */
public class Calendar extends SimplePanel implements Display {

  @UiTemplate("Calendar.ui.xml")
	public interface Binder extends UiBinder<Widget, Calendar> {
	}
	
	private static final Binder binder = GWT.create(Binder.class);

	private final CalendarCaption i_calendarCaption;
	private final CanWaitSupport i_canWaitSupport;
	
	@UiField DialogBox popupPanel;
	@UiField ListBox reminderHours;
	@UiField ListBox reminderMinutes;
	@UiField CheckBox busy;
	@UiField CheckBox share;
	@UiField CheckBox selected;
	@UiField Button changeColour;
	@UiField Button undo;
	@UiField Button cancel;
	@UiField Button save;
	@UiField Button remove;

	@Inject
	public Calendar(CalendarCaption calendarCaption, CanWaitSupport canWaitSupport) {
		i_calendarCaption = calendarCaption;
		add(binder.createAndBindUi(this));
		i_canWaitSupport = canWaitSupport;
		canWaitSupport.wrap(
		    getReminderHours(), getReminderMinutes(), getBusy(), getShare(), getSelected(), 
		    getChangeColour(), getUndo(), getCancel(), getSave(), getRemove());
	}

	@UiFactory
	public DialogBox createDialogBox(boolean animationEnabled, boolean glassEnabled) {
		DialogBox dialogBox = new DialogBox(getCalendarCaption());
		dialogBox.setAnimationEnabled(animationEnabled);
		dialogBox.setGlassEnabled(glassEnabled);
		return dialogBox;
	}
	
	@Override
	public void startWaiting() {
	  getCanWaitSupport().startWaiting();
	}

	@Override
	public void stopWaiting() {
	  getCanWaitSupport().stopWaiting();
	}
	
	@Override
	public CheckBox getBusy() {
		return busy;
	}

	@Override
	public CheckBox getShare() {
		return share;
	}

	@Override
	public CheckBox getSelected() {
		return selected;
	}

	@Override
	public Button getChangeColour() {
		return changeColour;
	}

	@Override
	public Button getUndo() {
		return undo;
	}

	@Override
	public Button getCancel() {
		return cancel;
	}
	@Override
	public Button getSave() {
		return save;
	}

	@Override
	public Button getRemove() {
		return remove;
	}

	public ListBox getReminderHours() {
		return reminderHours;
	}

	public ListBox getReminderMinutes() {
		return reminderMinutes;
	}

	public PopupPanel getPopupPanel() {
		return popupPanel;
	}

	public CalendarCaption getCalendarCaption() {
		return i_calendarCaption;
	}

  public CanWaitSupport getCanWaitSupport() {
    return i_canWaitSupport;
  }
}

File: app/controllers/Calendar.scala
/**
 * Copyright 2013 Alex Jones
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with work for additional information
 * regarding copyright ownership.  The ASF licenses file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package controllers

import java.io.StringWriter
import javax.inject.Inject
import play.api.mvc.Action
import uk.co.unclealex.hammers.calendar.cal.CalendarFactory
import uk.co.unclealex.hammers.calendar.cal.CalendarWriter
import uk.co.unclealex.hammers.calendar.search.AttendedSearchOption
import uk.co.unclealex.hammers.calendar.search.GameOrTicketSearchOption
import uk.co.unclealex.hammers.calendar.search.LocationSearchOption
import play.api.mvc.Controller
import javax.inject.Named
import uk.co.unclealex.hammers.calendar.update.LastUpdated

/**
 * The controller that handles generating calendars.
 * @author alex
 *
 */
class Calendar @Inject() (
  /**
   * The secret used to protect the update path.
   */
  @Named("secret") val secret: String,
  /**
   * The last update service to get when the calendars were last updated.
   */
  lastUpdated: LastUpdated,
  /**
   * The calendar factory used to generate calendars.
   */
  calendarFactory: CalendarFactory,
  /**
   * The calendar write used to write calendars.
   */
  calendarWriter: CalendarWriter) extends Controller with Secret with Etag {

  def searchSecure(secretPayload: String, attendedSearchOption: String, locationSearchOption: String, gameOrTicketSearchOption: String) =
    Secret(secretPayload) {
      calendar(
        None,
        AttendedSearchOption(attendedSearchOption),
        LocationSearchOption(locationSearchOption),
        GameOrTicketSearchOption(gameOrTicketSearchOption))
    }

  def search(mask: String, locationSearchOption: String, gameOrTicketSearchOption: String) =
    calendar(
      Some("busy" == mask),
      Some(AttendedSearchOption.ANY),
      LocationSearchOption(locationSearchOption),
      GameOrTicketSearchOption(gameOrTicketSearchOption))

  def calendar(
    busyMask: Option[Boolean],
    a: Option[AttendedSearchOption],
    l: Option[LocationSearchOption],
    g: Option[GameOrTicketSearchOption]) = {
    (a, l, g) match {
      case (Some(a), Some(l), Some(g)) => ETag(calculateETag(busyMask, Some(a), Some(l), Some(g))) {
        Action { implicit request =>
          val calendar = calendarFactory.create(busyMask, a, l, g)
          val buffer = new StringWriter
          calendarWriter.write(calendar, buffer)
          val output = buffer.toString
          Ok(output).as(calendarWriter.mimeType)
        }
      }
      case _ => Action { implicit request => NotFound }
    }
  }

  def calculateETag(parts: Option[Any]*): String = {
    val allParts = (lastUpdated when) :: parts.toList
    val toStringFactory = { any: Option[Any] => 
      any match {
        case Some(any) => any.toString
        case None => ""
      }
    }
    allParts.map(toStringFactory).mkString(",")
  }

}